### PR TITLE
add tensor clone

### PR DESCRIPTION
**Pull request by zhwesky2010**
### PR types

New features 
### PR changes

APIs
### Describe

1. 增加了了Tensor的clone功能，clone之后的Tensor与原来Tensor不共享数据，一定会发生Tensor拷贝，且place/shape/dtype与之前完全一致，并且提供梯度回溯功能；
```
import paddle

x = paddle.to_tensor(1.0, stop_gradient=False)
clone_x = x.clone()
y = clone_x**2
y.backward()
print(clone_x.stop_gradient) # False
print(clone_x.grad)          # [2.0], support gradient propagation
print(x.stop_gradient)       # False
print(x.grad)                # [2.0], clone_x support gradient propagation for x

x = paddle.to_tensor(1.0)
clone_x = x.clone()
clone_x.stop_gradient = False
z = clone_x**3
z.backward()
print(clone_x.stop_gradient) # False
print(clone_x.grad)          # [3.0], support gradient propagation
print(x.stop_gradient) # True
print(x.grad)          # None

```
2. 修改了assignOP的逻辑，分配out的Tensor时不再基于tracer的place，而是与input的Tensor位置保持一致，使得assign前后的place位置一致，因为2.0的Tensor是可以自由指定位置的，不一定与tracer相同；

与Tensor.detach的功能相关联，https://github.com/PaddlePaddle/Paddle/pull/27921

如果需要一个既分离计算图，又不共享数据内存的Tensor，请使用`x.detach().clone()`
```
import paddle
x = paddle.to_tensor(1.0, stop_gradient=False)
detach_clone_x = x.detach().clone()
```

### CODE REVIEW

**Review by phlrain — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM
